### ISSUE TITLE
[Bug?] Jax shuffle speed versus numpy random shuffle speed

### ISSUE THREAD

**Issue by zmjjmz** [google/jax#2301]
I was reexamining a training loop I'd made in jax for a talk, and noticed that when I tried to demonstrate that using `jax.random.shuffle` gives a speedup over using `numpy.random.shuffle`, I found that the latter was actually _significantly_ faster. I then went and recreated it in a small script, and I've found that the jax shuffle method gets really slow for ndarrays of around 1e6 and 1e7, both of which are sizes we easily encounter.

The demonstration code and speeds are in [this gist](https://gist.github.com/zmjjmz/6c9184971f8601dd923193a47640d651), but basically at around 1e7 `numpy.random.shuffle` takes a neat 3 seconds while `jax.random.shuffle` clocks in at over 4 minutes!

Is this intentional? Is it unavoidable due to not being able to do it in-place? I tried reading into the JAX shuffle implementation a bit and it seems like a lot of effort goes into making it rather fast, and the numpy random shuffle has the advantage of doing everything in place, so that would all make sense to me.

**Comment by shoyer** [comment]
I don't really know what's going on here, but this might simply be a short-coming of either XLA or this algorithm on CPUs. When I run your benchmark on CPUs, JAX is about 40x slower than NumPy, but on GPUs it's about 2x faster.

**Comment by zmjjmz** [comment]
Ah ok, yeah I can understand (especially given the warnings I get about using a CPU) that it's not the top development target. I haven't tested any of the stuff I've done on a GPU machine yet.

I'm assuming you mean that you ran the numpy portions on the CPU and JAX on the GPU?

**Comment by shoyer** [comment]
Yes, NumPy only runs on CPUs.

We could potentially switch to a different method on CPUs for JAX (could even make use of NumPy's implementation) but this would make JAX's random number generator platform specific. A safer option might be to add an option for switching the shuffle method.

**Comment by jakevdp** [comment]
Revisiting this: `jax.random.shuffle` has been deprecated in favor of `jax.random.permutation`, but the performance issue remains in the current main branch of JAX (run on a Colab CPU):
```python
import numpy as np
from jax import random
import jax.numpy as jnp
import jax

x = np.random.rand(1_000_000).astype('float32')

%timeit np.random.permutation(x)
# 10 loops, best of 5: 42.5 ms per loop

x_jax = jnp.asarray(x)
key = random.PRNGKey(7548923)
_ = random.permutation(key, x_jax)
%timeit random.permutation(key, x_jax).block_until_ready()
# 1 loop, best of 5: 1.24 s per loop

# routine is already JIT-compiled, so another JIT doesn't help much:
perm_jit = jax.jit(random.permutation)
_ = perm_jit(key, x_jax)
%timeit perm_jit(key, x_jax).block_until_ready()
# 1 loop, best of 5: 1.19 s per loop
```
On GPU, JAX looks better:
```
10 loops, best of 5: 57.6 ms per loop
10 loop, best of 5: 16.4 ms per loop
100 loops, best of 5: 10.6 ms per loop
```

I'm going to assign to @froystig as he's been working on PRNG updates and may have ideas about how to improve performance.

**Comment by akbir** [comment]
Keen to hear if there is any update on this.
(Happy to also help if directed!!)
